decision_service/include/ds_config_collection.h

```diff
@@ -0,0 +1,19 @@
+#pragma once
+#include <string>
+#include <unordered_map>
+
+namespace decision_service { namespace utility {
+
+  class config_collection
+  {
+  public:
+    void set(const char* name, const char* value);
+
+    const char* get(const char* name, const char* defval) const;
+    int get_int(const char* name, int defval) const;
+
+  private:
+    std::unordered_map<std::string, std::string> _map;
```

Comment:
What is the motivation for avoiding stl in the interface?

Reply:
I got burned in the past with linking against different versions of STL (e.g. VS 2013 vs 2015). if the internal structure you get very hard to debug crashes

---
In reply to: [186772707](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186772707) [](ancestors = 186772707)

Reply:
Confirming my understanding of the issue:
  
1) Inside api-consumer code if new config_collection() is called or a config_collection is created on the stack, the compiler will look at this header to determine how much space to allocate.
2) It will then call the constructor for config_collection which is in dsclientlib.  
3) config_collection constructor which is compiled in to config_collection.obj will call the constructor for std::unordered_map is also compiled into config_collection.obj
4) So the concern is that if unordered_map header included in api-consumer code has different fields, it might allocate more or less memory than needed.  If it is more, it should not cause a big issue but if is less, then we might run into a large issue.
5) This assumes that ds-clientlib is included as a pre-compiled binary

Is this correct?  Is there something else that worries you?

---
In reply to: [186815209](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186815209) [](ancestors = 186815209,186772707)